dali/operators/reader/loader/video/frames_decoder_test.cc

```diff
@@ -16,223 +16,219 @@
+    ground_truth.CompareFrame(0, frame.data());
 
     // Seek to frame
     file.SeekFrame(25);
     ASSERT_EQ(file.NextFrameIdx(), 25);
     file.ReadNextFrame(frame.data());
-    this->CompareFrames(frame.data(), this->GetVfrFrame(1, 25), file.FrameSize());
+    ground_truth.CompareFrame(25, frame.data());
 
     // Seek back to frame
     file.SeekFrame(12);
     ASSERT_EQ(file.NextFrameIdx(), 12);
     file.ReadNextFrame(frame.data());
-    this->CompareFrames(frame.data(), this->GetVfrFrame(1, 12), file.FrameSize());
+    ground_truth.CompareFrame(12, frame.data());
 
     // Seek to last frame (flush frame)
-    file.SeekFrame(59);
-    ASSERT_EQ(file.NextFrameIdx(), 59);
+    int last_frame_index = ground_truth.NumFrames() - 1;
+    file.SeekFrame(last_frame_index);
+    ASSERT_EQ(file.NextFrameIdx(), last_frame_index);
     file.ReadNextFrame(frame.data());
-    this->CompareFrames(frame.data(), this->GetVfrFrame(1, 59), file.FrameSize());
+    ground_truth.CompareFrame(last_frame_index, frame.data());
+    ASSERT_EQ(file.NextFrameIdx(), -1);
 
     // Wrap around to first frame
     ASSERT_FALSE(file.ReadNextFrame(frame.data()));
     file.Reset();
     ASSERT_EQ(file.NextFrameIdx(), 0);
     file.ReadNextFrame(frame.data());
-    this->CompareFrames(frame.data(), this->GetVfrFrame(1, 0), file.FrameSize());
-}
-
-TEST_F(FramesDecoderTest_CpuOnlyTests, InvalidPath) {
-    std::string path = "invalid_path.mp4";
-
-    try {
-        FramesDecoder file(path);
-    } catch (const DALIException &e) {
-        EXPECT_TRUE(strstr(
-            e.what(),
-            make_string("Failed to open video file at path ", path).c_str()));
-    }
-}
+    ground_truth.CompareFrame(0, frame.data());
 
-TEST_F(FramesDecoderTest_CpuOnlyTests, NoVideoStream) {
-    std::string path = testing::dali_extra_path() + "/db/audio/wav/dziendobry.wav";
-
-    try {
-        FramesDecoder file(path);
-    } catch (const DALIException &e) {
-        EXPECT_TRUE(strstr(
-            e.what(),
-            make_string("Could not find a valid video stream in a file ", path).c_str()));
-    }
-}
+    // Seek to random frames and read them
+    std::mt19937 gen(0);
+    std::uniform_int_distribution<> distr(0, last_frame_index);
 
-TEST_F(FramesDecoderTest_CpuOnlyTests, InvalidSeek) {
-    std::string path = testing::dali_extra_path() + "/db/video/cfr/test_1.mp4";
-    FramesDecoder file(path);
+    for (int i = 0; i < 20; ++i) {
+      int next_index = distr(gen);
 
-    try {
-        file.SeekFrame(60);
-    } catch (const DALIException &e) {
-        EXPECT_TRUE(strstr(
-            e.what(),
-            "Invalid seek frame id. frame_id = 60, num_frames = 50"));
+      file.SeekFrame(next_index);
+      file.ReadNextFrame(frame.data());
+      ground_truth.CompareFrame(next_index, frame.data());
     }
-}
+  }
+};
 
-TEST_F(FramesDecoderGpuTest, VariableFrameRateGpu) {
+class FramesDecoderGpuTest : public FramesDecoderTestBase {
```

Comment:
Is it possible to unify this with FramesDecoderTest_CpuOnlyTests?
The main difference I see is the additional copy of the data from `DeviceBuffer` to `frame_cpu`, and the different method of validation compared to `ground_truth`. I think these two things could be just a specialization of a method for the GPU test, but other steps could be common (the flow of the tests itself).

Reply:
Didn't want to over do it in the beginning but sure, done.